File: server/src/entities/memory.entity.ts
import { AssetEntity } from 'src/entities/asset.entity';
import { UserEntity } from 'src/entities/user.entity';
import { MemoryType } from 'src/enum';
import {
  Column,
  CreateDateColumn,
  DeleteDateColumn,
  Entity,
  Index,
  JoinTable,
  ManyToMany,
  ManyToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';

export type OnThisDayData = { year: number };

export interface MemoryData {
  [MemoryType.ON_THIS_DAY]: OnThisDayData;
}

@Entity('memories')
export class MemoryEntity<T extends MemoryType = MemoryType> {
  @PrimaryGeneratedColumn('uuid')
  id!: string;

  @CreateDateColumn({ type: 'timestamptz' })
  createdAt!: Date;

  @UpdateDateColumn({ type: 'timestamptz' })
  updatedAt!: Date;

  @Index('IDX_memories_update_id')
  @Column({ type: 'uuid', nullable: false, default: () => 'immich_uuid_v7()' })
  updateId?: string;

  @DeleteDateColumn({ type: 'timestamptz' })
  deletedAt?: Date;

  @ManyToOne(() => UserEntity, { onDelete: 'CASCADE', onUpdate: 'CASCADE', nullable: false })
  owner!: UserEntity;

  @Column()
  ownerId!: string;

  @Column()
  type!: T;

  @Column({ type: 'jsonb' })
  data!: MemoryData[T];

  /** unless set to true, will be automatically deleted in the future */
  @Column({ default: false })
  isSaved!: boolean;

  /** memories are sorted in ascending order by this value */
  @Column({ type: 'timestamptz' })
  memoryAt!: Date;

  @Column({ type: 'timestamptz', nullable: true })
  showAt?: Date;

  @Column({ type: 'timestamptz', nullable: true })
  hideAt?: Date;

  /** when the user last viewed the memory */
  @Column({ type: 'timestamptz', nullable: true })
  seenAt?: Date;

  @ManyToMany(() => AssetEntity)
  @JoinTable()
  assets!: AssetEntity[];
}

File: mobile/lib/infrastructure/entities/memory.entity.dart
import 'package:drift/drift.dart';
import 'package:immich_mobile/domain/models/memory.model.dart';
import 'package:immich_mobile/infrastructure/entities/user.entity.dart';
import 'package:immich_mobile/infrastructure/utils/datetime_clamp.type.dart';
import 'package:immich_mobile/infrastructure/utils/drift_default.mixin.dart';

class MemoryEntity extends Table with DriftDefaultsMixin {
  const MemoryEntity();

  TextColumn get id => text()();

  DateTimeColumn get createdAt => customType(clampedDateTime).withDefault(currentDateAndTime)();

  DateTimeColumn get updatedAt => customType(clampedDateTime).withDefault(currentDateAndTime)();

  DateTimeColumn get deletedAt => customType(clampedDateTime).nullable()();

  TextColumn get ownerId => text().references(UserEntity, #id, onDelete: KeyAction.cascade)();

  IntColumn get type => intEnum<MemoryTypeEnum>()();

  TextColumn get data => text()();

  BoolColumn get isSaved => boolean().withDefault(const Constant(false))();

  DateTimeColumn get memoryAt => customType(clampedDateTime)();

  DateTimeColumn get seenAt => customType(clampedDateTime).nullable()();

  DateTimeColumn get showAt => customType(clampedDateTime).nullable()();

  DateTimeColumn get hideAt => customType(clampedDateTime).nullable()();

  @override
  Set<Column> get primaryKey => {id};
}
